### PR TITLE
Make Mantid work on windows with new setup tools versions

### LINKED ISSUE #33118: Mantid doesn't work with latest setuptools on Windows
I recently updated the Python setuptools package on my Windows laptop by accident and I noticed that workbench and system tests stopped working. I looked into this a bit and came across some issues where support had been added to Mantid for setuptools v49.0.0 or later (eg https://github.com/mantidproject/mantid/pull/32068).

The problem on Windows seems to be that the python version is 3.8 which includes a new importlib.metadata module. This alters the behaviour of the shim (workbench-script.pyw) created by setuptools v49.0.0 or later:

```
#!C:\mantid\external\src\ThirdParty\lib\python3.8\pythonw.exe
# EASY-INSTALL-ENTRY-SCRIPT: 'MantidWorkbench','gui_scripts','workbench'
import re
import sys

# for compatibility with easy_install; see #2198
__requires__ = 'MantidWorkbench'

try:
    from importlib.metadata import distribution
except ImportError:
    try:
        from importlib_metadata import distribution
    except ImportError:
        from pkg_resources import load_entry_point


def importlib_load_entry_point(spec, group, name):
    dist_name, _, _ = spec.partition('==')
    matches = (
        entry_point
        for entry_point in distribution(dist_name).entry_points
        if entry_point.group == group and entry_point.name == name
    )
    return next(matches).load()


globals().setdefault('load_entry_point', importlib_load_entry_point)


if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(load_entry_point('MantidWorkbench', 'gui_scripts', 'workbench')())

```

Ubuntu uses python 3.6 so the shim imports pkg_resources instead.

This difference matters because it appears to me that importlib doesn't recognise .egg-link files whereas pkg_resources does. The .egg-link file needs to be read in order to locate the MantidWorkbench package in a developer's source code directory at qt\applications\workbench

Mantid works OK on older versions of setuptools because the shim is different to the one above and always uses pkg_resources

Maybe this isn't A1 priority but I wondered about fixing it because it wasn't v obvious what was going wrong - I just had a dev environment which wouldn't work and failed with a "cannot find _kernel.." error*. Not sure what can be done to fix this though - perhaps there's a way of getting setuptools to create a custom .pyw file somehow. The standard .pyw file seems to be autogenerated when you use the gui_scripts entry point in the qt\applications\workbench\setup.py file.

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The setuptools develop mode previously allowed workbench to run successfully by providing .egg-link files such as MantidWorkbench.egg-link in the `bin\<config>` directory which helped locate python packages in the source code directories. The .egg-link files aren't supported any more if using python 3.8 and setuptools>=49.0.0 so an alternative mechanism is needed.

This PR sets up two sitecustomize.py files that sort this out:
a) a sitecustomize.py file in the `bin/<config>` directory that sets up `bin/<config>` as a site directory. This causes the .pth files in here to be read and achieve the same thing as the .egg-link files
b) a sitecustomize.py file in the python site-packages directory to ensure that the file in a) is found. The cwd is added to the path

Notes:
The file in a) has actually been called sitecustomize_mantid.py to ensure both files are found and one doesn't hide the other. Python only supports finding x1 sitecustomize.py file and the search order through site-packages and `bin/<config>` directory isn't the same for all methods of starting up workbench

There was already some cmake logic to create the file in a) but I don't think it was being used in most situations. The .egg-link files were being relied upon instead.

The sitecustomize.py file in b) already existed for Windows users and was used to set up some paths to the 3rd party libraries. It was stored in the third-partymsvc repository. Since it's needed for all users with python3.8 and setuptools>=49.0.0 it makes more sense to create it in cmake

**To test:**

I think this needs testing on various platforms so may leave this until the next maintenance sprint:

Upgrade setuptools to a new version: `pip install "setuptools==49.0.0"`
Make a note of the original setuptools version reported in the message so you can switch back at the end
Run cmake
Build the workbench target which will generate a new shim file in the `bin/<config>` directory. This is called workbench-script.pyw on Windows
Try launching workbench using the following methods:
a) from PyCharm using the Run, Run and Run, Debug methods
b) On Windows, using the MantidWorkbench target

Fixes #33118 . 

*This does not require release notes* because **there's no functional change to users**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import os
+sys.path = [os.getcwd()] + sys.path
+try:
+    import sitecustomize_mantid
+except ImportError:
+    pass
+"
+  )
+
+  file(WRITE ${PYTHON_SITE_PACKAGES}/sitecustomize.py "${SITE_CUSTOMIZE_CONTENT}")
```
Comment (martyngigg) [inline buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake:93]: Just so I understand the process - is this essentially here to get the `cwd` onto the `sys.path`?

Could I ask where this goes on Linux? The `site-packages` from the system would not be writable I don't think?

File: buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    # We assume the layout of the thirdparty directory has DLLs in bin, lib\\qt4 and lib\\qt5
+    thirdparty_lib = os.path.dirname(py_exe_dir)
+    thirdparty_bin = os.path.dirname(thirdparty_lib) + \"\\\\bin\"
+
+    extra_paths = (
+        py_exe_dir,  # required for C++ tests that embed Python to find Python DLL
+        thirdparty_bin,
+        thirdparty_bin + \"\\\\mingw\",
+        thirdparty_lib + \"\\\\qt4\\\\bin\",
```
Comment (martyngigg) [inline buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake:49]: Can we remove the Qt4 ones from here?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Will we need to remove the `sitecustomize.py` in the third party repo now?

File: buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake (hunk 1 above)
Comment (DannyHindson) [reply buildconfig/CMake/WriteSiteCustomizeToSitePackages.cmake:93]: Yes the content of this sitecustomize.py file that applies to all platforms is aimed at getting the `cwd` onto the sys.path. If `site-packages` is not writable on some platforms then that is a problem :(

**Comment by martyngigg** [comment]
@DannyHindson I'm expecting the conda switch might make this less of a requirement. Should we close this and revisit after this release?

**Comment by DannyHindson** [comment]
> @DannyHindson I'm expecting the conda switch might make this less of a requirement. Should we close this and revisit after this release?

Yes that's fine. I think I'd got a bit stuck with this given that my most recent solution was based on `site-packages` being writable

**Pull request closed without merging**